File: src/agcoord/resources.py
"""Backend-neutral contracts for admitted and enforced machine resources.

Resource names remain project-defined admission tokens until the broker configuration
binds a name to a typed unit and an enforcement backend.  This module deliberately owns
only the contract and backend lifecycle seam; concrete cgroup, filesystem, and container
implementations live behind that seam.
"""

from __future__ import annotations

from dataclasses import dataclass
import json
import os
import re
from types import MappingProxyType
from typing import Mapping, Protocol, runtime_checkable


RESOURCE_MODES = ("admission-only", "best-effort", "required")
RESOURCE_UNITS_BY_KIND = {
    "generic": frozenset({"admission-unit"}),
    "cpu": frozenset({"logical-cpu"}),
    "memory": frozenset({"bytes"}),
    "memory-high": frozenset({"bytes"}),
    "swap": frozenset({"bytes"}),
    "tmpfs": frozenset({"bytes"}),
    "storage": frozenset({"bytes"}),
    "io-bandwidth": frozenset({"bytes-per-second"}),
    "io-operations": frozenset({"operations-per-second"}),
    "inodes": frozenset({"inodes"}),
    "processes": frozenset({"processes"}),
}
RESOURCE_OPERATIONS = (
    "prepare",
    "attach",
    "usage",
    "finish",
    "cancel",
    "cleanup",
)
RESOURCE_STAGES = frozenset(("probe", *RESOURCE_OPERATIONS))
RESOURCE_EVENT_STATUSES = frozenset({"applied", "recorded", "unapplied", "failed"})
_NAME = re.compile(r"^[a-z][a-z0-9_.:-]{0,63}$")
_CODE = re.compile(r"^[a-z][a-z0-9-]{0,63}$")
_BINDING_KEYS = frozenset({"kind", "unit", "mode", "backend"})
_CAPABILITY_KEYS = frozenset(
    {"available", "kinds", "units", "operations", "reason"}
)
_RECEIPT_KEYS = frozenset({"requested", "applied", "peak", "events"})
_EVENT_KEYS = frozenset({"at", "backend", "resource", "stage", "status", "code"})
ADMISSION_BINDING: dict[str, object] = {
    "backend": None,
    "kind": "generic",
    "mode": "admission-only",
    "unit": "admission-unit",
}


class ResourceContractError(RuntimeError):
    """A malformed public resource contract or backend response."""


class ResourceBackendError(ResourceContractError):
    """A backend refusal with a sanitized code safe for durable public receipts."""

    def __init__(self, code: str, message: str) -> None:
        if not isinstance(code, str) or not _CODE.fullmatch(code):
            raise ValueError("resource backend error code is invalid")
        super().__init__(message)
        self.code = code


@dataclass(frozen=True)
class ResourceObservation:
    """One sanitized backend fact to record in the durable resource events."""

    resource: str
    code: str


@dataclass(frozen=True)
class ResourceMeasurement:
    """Peak values plus stable, non-sensitive observations from one sample."""

    peak: Mapping[str, int]
    observations: tuple[ResourceObservation, ...] = ()


@dataclass(frozen=True)
class ResourceRequest:
    """One backend's immutable subset of an admitted run's resource contract."""

    run_id: str
    backend: str
    resources: Mapping[str, int]
    bindings: Mapping[str, Mapping[str, object]]

    @classmethod
    def build(
        cls,
        run_id: str,
        backend: str,
        resources: Mapping[str, int],
        bindings: Mapping[str, Mapping[str, object]],
    ) -> ResourceRequest:
        frozen_bindings = {
            name: MappingProxyType(dict(binding))
            for name, binding in sorted(bindings.items())
        }
        return cls(
            run_id=run_id,
            backend=backend,
            resources=MappingProxyType(dict(sorted(resources.items()))),
            bindings=MappingProxyType(frozen_bindings),
        )


@runtime_checkable
class ResourceBackend(Protocol):
    """Lifecycle implemented by an optional native or container resource backend.

    ``prepare`` must return JSON-serializable private state. ``attach`` applies the
    prepared controls before AGCoord releases the blocked worker launcher. Lifecycle
    methods must be idempotent because a replacement broker can resume a durable row.
    Public failures are converted to stable event codes instead of exposing backend
    exception text or host paths.
    """

    def probe(self) -> Mapping[str, object]: ...

    def prepare(self, request: ResourceRequest) -> Mapping[str, object]: ...

    def attach(
        self,
        request: ResourceRequest,
        state: Mapping[str, object],
        worker_pid: int,
    ) -> None: ...

    def usage(
        self,
        request: ResourceRequest,
        state: Mapping[str, object],
    ) -> Mapping[str, int] | ResourceMeasurement: ...

    def finish(
        self,
        request: ResourceRequest,
        state: Mapping[str, object],
    ) -> Mapping[str, int] | ResourceMeasurement: ...

    def cancel(
        self,
        request: ResourceRequest,
        state: Mapping[str, object],
    ) -> None: ...

    def cleanup(
        self,
        request: ResourceRequest,
        state: Mapping[str, object],
    ) -> None: ...


def _resource_mapping(
    value: object,
    *,
    subject: str,
    allow_zero: bool,
) -> dict[str, int]:
    if not isinstance(value, Mapping):
        raise ResourceContractError(f"{subject} must be a resource-to-integer mapping")
    selected: dict[str, int] = {}
    for name, units in value.items():
        if not isinstance(name, str) or not _NAME.fullmatch(name):
            raise ResourceContractError(f"invalid {subject} resource name {name!r}")
        if (
            not isinstance(units, int)
            or isinstance(units, bool)
            or units < 0
            or (not allow_zero and units == 0)
        ):
            qualifier = "non-negative" if allow_zero else "positive"
            raise ResourceContractError(
                f"{subject} resource {name!r} must be a {qualifier} integer"
            )
        selected[name] = units
    return dict(sorted(selected.items()))


def validate_resource_bindings(
    value: Mapping[str, Mapping[str, object]] | None,
) -> dict[str, dict[str, object]]:
    """Normalize strict machine bindings without making familiar names special."""
    if value is None:
        return {}
    if not isinstance(value, Mapping):
        raise ResourceContractError("resource bindings must be a name-to-binding mapping")
    selected: dict[str, dict[str, object]] = {}
    for name, raw in value.items():
        if not isinstance(name, str) or not _NAME.fullmatch(name):
            raise ResourceContractError(f"invalid resource binding name {name!r}")
        if not isinstance(raw, Mapping) or set(raw) != _BINDING_KEYS:
            raise ResourceContractError(
                f"resource binding {name!r} must contain exactly "
                "backend, kind, mode, and unit"
            )
        kind = raw["kind"]
        unit = raw["unit"]
        mode = raw["mode"]
        backend = raw["backend"]
        if not isinstance(kind, str) or kind not in RESOURCE_UNITS_BY_KIND:
            raise ResourceContractError(
                f"resource binding {name!r} has unknown kind {kind!r}"
            )
        if not isinstance(unit, str) or unit not in RESOURCE_UNITS_BY_KIND[kind]:
            raise ResourceContractError(
                f"resource binding {name!r} cannot use unit {unit!r} for kind {kind!r}"
            )
        if not isinstance(mode, str) or mode not in RESOURCE_MODES:
            raise ResourceContractError(
                f"resource binding {name!r} has unknown mode {mode!r}"
            )
        if mode == "admission-only":
            if backend is not None:
                raise ResourceContractError(
                    f"admission-only resource binding {name!r} cannot select a backend"
                )
        elif not isinstance(backend, str) or not _NAME.fullmatch(backend):
            raise ResourceContractError(
                f"enforced resource binding {name!r} needs a valid backend name"
            )
        selected[name] = {
            "backend": backend,
            "kind": kind,
            "mode": mode,
            "unit": unit,
        }
    return dict(sorted(selected.items()))


def configured_resource_bindings(
    section: Mapping[str, Mapping[str, object]] | None,
) -> dict[str, dict[str, object]]:
    """Validate the bindings section of one state directory's configuration file."""
    if section is None:
        return {}
    return validate_resource_bindings(section)


def resource_contract(
    resources: Mapping[str, int],
    bindings: Mapping[str, Mapping[str, object]],
) -> dict[str, dict[str, object]]:
    requested = _resource_mapping(resources, subject="requested", allow_zero=False)
    configured = validate_resource_bindings(bindings)
    return {
        name: dict(configured.get(name, ADMISSION_BINDING))
        for name in requested
    }


def validate_resource_contract(
    value: object,
    resources: Mapping[str, int],
) -> dict[str, dict[str, object]]:
    requested = _resource_mapping(resources, subject="requested", allow_zero=False)
    contract = validate_resource_bindings(value if isinstance(value, Mapping) else None)
    if set(contract) != set(requested):
        raise ResourceContractError(
            "resource contract names do not match the requested resource names"
        )
    return contract


def initial_resource_receipt(resources: Mapping[str, int]) -> dict[str, object]:
    return {
        "requested": _resource_mapping(
            resources,
            subject="requested",
            allow_zero=False,
        ),
        "applied": {},
        "peak": {},
        "events": [],
    }


def validate_resource_measurement(
    value: object,
    *,
    expected: set[str],
) -> tuple[dict[str, int], tuple[ResourceObservation, ...]]:
    """Normalize one backend sample without admitting unexpected public names."""

    if isinstance(value, ResourceMeasurement):
        raw_peak = value.peak
        raw_observations = value.observations
    else:
        raw_peak = value
        raw_observations = ()
    peak = _resource_mapping(
        raw_peak,
        subject="resource usage",
        allow_zero=True,
    )
    if not set(peak).issubset(expected):
        raise ResourceContractError("resource usage names an unexpected resource")
    if not isinstance(raw_observations, tuple):
        raise ResourceContractError("resource observations must be a tuple")
    observations: list[ResourceObservation] = []
    for observation in raw_observations:
        if not isinstance(observation, ResourceObservation):
            raise ResourceContractError("resource observation is invalid")
        if (
            not isinstance(observation.resource, str)
            or observation.resource not in expected
        ):
            raise ResourceContractError(
                "resource observation names an unexpected resource"
            )
        if not isinstance(observation.code, str) or not _CODE.fullmatch(
            observation.code
        ):
            raise ResourceContractError("resource observation code is invalid")
        if observation not in observations:
            observations.append(observation)
    return peak, tuple(observations)


def validate_resource_receipt(
    value: object,
    resources: Mapping[str, int],
) -> dict[str, object]:
    requested_resources = _resource_mapping(
        resources,
        subject="requested",
        allow_zero=False,
    )
    if not isinstance(value, Mapping) or set(value) != _RECEIPT_KEYS:
        raise ResourceContractError(
            "resource receipt must contain exactly requested, applied, peak, and events"
        )
    requested = _resource_mapping(
        value["requested"],
        subject="receipt requested",
        allow_zero=False,
    )
    if requested != requested_resources:
        raise ResourceContractError("resource receipt request does not match the run")
    applied = _resource_mapping(
        value["applied"],
        subject="receipt applied",
        allow_zero=False,
    )
    peak = _resource_mapping(
        value["peak"],
        subject="receipt peak",
        allow_zero=True,
    )
    if not set(applied).issubset(requested):
        raise ResourceContractError("resource receipt applied unknown resources")
    if any(applied[name] > requested[name] for name in applied):
        raise ResourceContractError("resource receipt applied more than was requested")
    if not set(peak).issubset(requested):
        raise ResourceContractError("resource receipt measured unknown resources")
    raw_events = value["events"]
    if not isinstance(raw_events, list):
        raise ResourceContractError("resource receipt events must be a list")
    events: list[dict[str, str]] = []
    for raw in raw_events:
        if not isinstance(raw, Mapping) or set(raw) != _EVENT_KEYS:
            raise ResourceContractError(
                "resource receipt events have a strict six-field shape"
            )
        at = raw["at"]
        backend = raw["backend"]
        resource = raw["resource"]
        stage = raw["stage"]
        status = raw["status"]
        code = raw["code"]
        if not isinstance(at, str) or not at:
            raise ResourceContractError("resource event timestamp must be non-empty")
        if not isinstance(backend, str) or not _NAME.fullmatch(backend):
            raise ResourceContractError("resource event backend is invalid")
        if resource not in requested:
            raise ResourceContractError("resource event names an unrequested resource")
        if stage not in RESOURCE_STAGES:
            raise ResourceContractError(f"resource event stage {stage!r} is invalid")
        if status not in RESOURCE_EVENT_STATUSES:
            raise ResourceContractError(f"resource event status {status!r} is invalid")
        if not isinstance(code, str) or not _CODE.fullmatch(code):
            raise ResourceContractError("resource event code is invalid")
        events.append(
            {
                "at": at,
                "backend": backend,
                "resource": resource,
                "stage": stage,
                "status": status,
                "code": code,
            }
        )
    return {
        "requested": requested,
        "applied": applied,
        "peak": peak,
        "events": events,
    }


def validate_resource_backends(
    value: Mapping[str, ResourceBackend] | None,
) -> dict[str, ResourceBackend]:
    if value is None:
        return {}
    if not isinstance(value, Mapping):
        raise ResourceContractError("resource backends must be a name-to-backend mapping")
    selected: dict[str, ResourceBackend] = {}
    for name, backend in value.items():
        if not isinstance(name, str) or not _NAME.fullmatch(name):
            raise ResourceContractError(f"invalid resource backend name {name!r}")
        if not isinstance(backend, ResourceBackend):
            raise ResourceContractError(
                f"resource backend {name!r} does not implement the complete lifecycle"
            )
        selected[name] = backend
    return dict(sorted(selected.items()))


def _capability(value: object, *, backend: str) -> dict[str, object]:
    if not isinstance(value, Mapping) or set(value) != _CAPABILITY_KEYS:
        raise ResourceContractError(
            f"resource backend {backend!r} returned an invalid capability shape"
        )
    available = value["available"]
    if not isinstance(available, bool):
        raise ResourceContractError(
            f"resource backend {backend!r} capability availability must be boolean"
        )

    def names(field: str, allowed: set[str] | frozenset[str]) -> list[str]:
        raw = value[field]
        if not isinstance(raw, list) or not all(isinstance(item, str) for item in raw):
            raise ResourceContractError(
                f"resource backend {backend!r} capability {field} must be a string list"
            )
        if len(raw) != len(set(raw)) or any(item not in allowed for item in raw):
            raise ResourceContractError(
                f"resource backend {backend!r} capability {field} is invalid"
            )
        return sorted(raw)

    kinds = names("kinds", set(RESOURCE_UNITS_BY_KIND))
    units = names(
        "units",
        frozenset().union(*RESOURCE_UNITS_BY_KIND.values()),
    )
    operations = names("operations", set(RESOURCE_OPERATIONS))
    reason = value["reason"]
    if available:
        if reason is not None:
            raise ResourceContractError(
                f"available resource backend {backend!r} cannot have a refusal reason"
            )
    elif not isinstance(reason, str) or not _CODE.fullmatch(reason):
        raise ResourceContractError(
            f"unavailable resource backend {backend!r} needs a stable reason code"
        )
    return {
        "available": available,
        "kinds": kinds,
        "units": units,
        "operations": operations,
        "reason": reason,
    }


def probe_resource_backends(
    backends: Mapping[str, ResourceBackend],
    bindings: Mapping[str, Mapping[str, object]],
) -> dict[str, dict[str, object]]:
    selected_backends = validate_resource_backends(backends)
    selected_bindings = validate_resource_bindings(bindings)
    referenced = {
        binding["backend"]
        for binding in selected_bindings.values()
        if binding["backend"] is not None
    }
    capabilities: dict[str, dict[str, object]] = {}
    for name in sorted(set(selected_backends) | referenced):
        backend = selected_backends.get(name)
        if backend is None:
            capabilities[name] = {
                "available": False,
                "kinds": [],
                "units": [],
                "operations": [],
                "reason": "backend-unavailable",
            }
            continue
        try:
            capabilities[name] = _capability(backend.probe(), backend=name)
        except Exception:
            capabilities[name] = {
                "available": False,
                "kinds": [],
                "units": [],
                "operations": [],
                "reason": "probe-failed",
            }
    return capabilities


def validate_resource_capabilities(
    value: object,
) -> dict[str, dict[str, object]]:
    if not isinstance(value, Mapping):
        raise ResourceContractError("resource capabilities must be a backend mapping")
    selected: dict[str, dict[str, object]] = {}
    for name, raw in value.items():
        if not isinstance(name, str) or not _NAME.fullmatch(name):
            raise ResourceContractError(f"invalid resource capability backend {name!r}")
        selected[name] = _capability(raw, backend=name)
    return dict(sorted(selected.items()))


def capability_issue(
    binding: Mapping[str, object],
    capability: Mapping[str, object] | None,
) -> str | None:
    if capability is None:
        return "backend-unavailable"
    if not capability["available"]:
        return str(capability["reason"])
    if binding["kind"] not in capability["kinds"]:
        return "kind-unsupported"
    if binding["unit"] not in capability["units"]:
        return "unit-unsupported"
    if not set(RESOURCE_OPERATIONS).issubset(capability["operations"]):
        return "lifecycle-unsupported"
    return None


def validate_backend_state(value: object) -> dict[str, dict[str, object]]:
    """Validate private durable handles without exposing their values publicly."""
    if not isinstance(value, Mapping):
        raise ResourceContractError("resource backend state must be a backend mapping")
    selected: dict[str, dict[str, object]] = {}
    for backend, raw in value.items():
        if not isinstance(backend, str) or not _NAME.fullmatch(backend):
            raise ResourceContractError("resource backend state has an invalid name")
        if not isinstance(raw, Mapping) or set(raw) != {
            "handle",
            "resources",
            "finished",
            "cancelled",
        }:
            raise ResourceContractError(
                f"resource backend state {backend!r} has an invalid shape"
            )
        if not isinstance(raw["finished"], bool) or not isinstance(raw["cancelled"], bool):
            raise ResourceContractError(
                f"resource backend state {backend!r} has invalid lifecycle flags"
            )
        resources = raw["resources"]
        if (
            not isinstance(resources, list)
            or not resources
            or len(resources) != len(set(resources))
            or not all(isinstance(name, str) and _NAME.fullmatch(name) for name in resources)
        ):
            raise ResourceContractError(
                f"resource backend state {backend!r} has invalid resource names"
            )
        try:
            handle = json.loads(json.dumps(raw["handle"], separators=(",", ":")))
        except (TypeError, ValueError) as exc:
            raise ResourceContractError(
                f"resource backend {backend!r} returned non-JSON private state"
            ) from exc
        if not isinstance(handle, dict):
            raise ResourceContractError(
                f"resource backend {backend!r} private state must be a JSON object"
            )
        selected[backend] = {
            "handle": handle,
            "resources": sorted(resources),
            "finished": raw["finished"],
            "cancelled": raw["cancelled"],
        }
    return dict(sorted(selected.items()))


def resource_enforcement_summary(receipt: Mapping[str, object]) -> str:
    """Return one stable operator label derived from a strict durable receipt."""
    events = receipt.get("events", [])
    applied = bool(receipt.get("applied", {}))
    selected_events = (
        [event for event in events if isinstance(event, Mapping)]
        if isinstance(events, list)
        else []
    )
    failed = any(event.get("status") == "failed" for event in selected_events)
    unapplied = any(event.get("status") == "unapplied" for event in selected_events)
    if failed:
        return "failed"
    if applied and unapplied:
        return "partial"
    if unapplied:
        return "unapplied"
    if applied:
        return "applied"
    return "admission-only"

File: native/agcoord-broker/src/resources.rs
use crate::error::{AppError, Result};
use serde_json::{Map, Value, json};
use std::collections::{BTreeMap, BTreeSet};
use std::fs;
use std::path::{Path, PathBuf};

pub const CGROUP_BACKEND: &str = "cgroup-v2";
pub const RESOURCE_OPERATIONS: [&str; 6] =
    ["prepare", "attach", "usage", "finish", "cancel", "cleanup"];

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct Binding {
    pub backend: Option<String>,
    pub kind: String,
    pub mode: String,
    pub unit: String,
}

impl Binding {
    pub fn admission() -> Self {
        Self {
            backend: None,
            kind: "generic".to_owned(),
            mode: "admission-only".to_owned(),
            unit: "admission-unit".to_owned(),
        }
    }

    pub fn enforced(&self) -> bool {
        self.mode != "admission-only"
    }

    pub fn required(&self) -> bool {
        self.mode == "required"
    }

    pub fn to_value(&self) -> Value {
        json!({
            "backend": self.backend,
            "kind": self.kind,
            "mode": self.mode,
            "unit": self.unit,
        })
    }
}

#[derive(Clone, Debug)]
pub struct ResourceConfiguration {
    pub capacities: BTreeMap<String, u64>,
    pub bindings: BTreeMap<String, Binding>,
    pub cgroup_root: Option<PathBuf>,
    pub cgroup_io_paths: Vec<PathBuf>,
}

#[derive(Clone, Debug)]
pub struct Capability {
    pub available: bool,
    pub kinds: BTreeSet<String>,
    pub units: BTreeSet<String>,
    pub operations: BTreeSet<String>,
    pub reason: Option<String>,
}

#[derive(Clone, Debug)]
pub struct BackendState {
    pub handle: Value,
    pub resources: Vec<String>,
    pub finished: bool,
    pub cancelled: bool,
}

pub fn parse_backend_state(value: &Value) -> Option<BTreeMap<String, BackendState>> {
    let object = value.as_object()?;
    let expected = BTreeSet::from(["cancelled", "finished", "handle", "resources"]);
    let mut selected = BTreeMap::new();
    let mut claimed = BTreeSet::new();
    for (backend, raw) in object {
        if !name_valid(backend) {
            return None;
        }
        let raw = raw.as_object()?;
        if raw.keys().map(String::as_str).collect::<BTreeSet<_>>() != expected {
            return None;
        }
        let handle = raw.get("handle")?.as_object()?;
        let mut names = raw
            .get("resources")?
            .as_array()?
            .iter()
            .map(|name| {
                name.as_str()
                    .filter(|name| name_valid(name))
                    .map(str::to_owned)
            })
            .collect::<Option<Vec<_>>>()?;
        names.sort();
        if names.is_empty()
            || names.windows(2).any(|pair| pair[0] == pair[1])
            || names.iter().any(|name| !claimed.insert(name.clone()))
        {
            return None;
        }
        selected.insert(
            backend.clone(),
            BackendState {
                handle: Value::Object(handle.clone()),
                resources: names,
                finished: raw.get("finished")?.as_bool()?,
                cancelled: raw.get("cancelled")?.as_bool()?,
            },
        );
    }
    Some(selected)
}

pub fn backend_state_value(state: &BTreeMap<String, BackendState>) -> Value {
    Value::Object(
        state
            .iter()
            .map(|(backend, record)| {
                (
                    backend.clone(),
                    json!({
                        "handle": record.handle,
                        "resources": record.resources,
                        "finished": record.finished,
                        "cancelled": record.cancelled,
                    }),
                )
            })
            .collect(),
    )
}

impl Capability {
    pub fn unavailable(reason: &str) -> Self {
        Self {
            available: false,
            kinds: BTreeSet::new(),
            units: BTreeSet::new(),
            operations: BTreeSet::new(),
            reason: Some(reason.to_owned()),
        }
    }

    pub fn to_value(&self) -> Value {
        json!({
            "available": self.available,
            "kinds": self.kinds,
            "units": self.units,
            "operations": self.operations,
            "reason": self.reason,
        })
    }
}

fn config_error(message: impl Into<String>) -> AppError {
    AppError::new("broker-config-invalid", message)
}

pub fn name_valid(value: &str) -> bool {
    let bytes = value.as_bytes();
    !bytes.is_empty()
        && bytes.len() <= 64
        && bytes[0].is_ascii_lowercase()
        && bytes[1..].iter().all(|byte| {
            byte.is_ascii_lowercase()
                || byte.is_ascii_digit()
                || matches!(byte, b'_' | b'.' | b':' | b'-')
        })
}

pub fn code_valid(value: &str) -> bool {
    let bytes = value.as_bytes();
    !bytes.is_empty()
        && bytes.len() <= 64
        && bytes[0].is_ascii_lowercase()
        && bytes[1..]
            .iter()
            .all(|byte| byte.is_ascii_lowercase() || byte.is_ascii_digit() || *byte == b'-')
}

fn valid_kind_unit(kind: &str, unit: &str) -> bool {
    match kind {
        "generic" => unit == "admission-unit",
        "cpu" => unit == "logical-cpu",
        "memory" | "memory-high" | "swap" | "tmpfs" | "storage" => unit == "bytes",
        "io-bandwidth" => matches!(
            unit,
            "bytes-per-second" | "read-bytes-per-second" | "write-bytes-per-second"
        ),
        "io-operations" => matches!(
            unit,
            "operations-per-second" | "read-operations-per-second" | "write-operations-per-second"
        ),
        "io-weight" => unit == "weight",
        "inodes" => unit == "inodes",
        "processes" => unit == "processes",
        _ => false,
    }
}

pub fn parse_bindings(value: Option<&Value>) -> Result<BTreeMap<String, Binding>> {
    let Some(value) = value else {
        return Ok(BTreeMap::new());
    };
    let bindings = value
        .as_object()
        .ok_or_else(|| config_error("broker configuration bindings must be a JSON object"))?;
    let mut selected = BTreeMap::new();
    for (name, raw) in bindings {
        if !name_valid(name) {
            return Err(config_error("resource binding name is invalid"));
        }
        let raw = raw.as_object().ok_or_else(|| {
            config_error(format!("resource binding {name} must be a JSON object"))
        })?;
        let expected = BTreeSet::from(["backend", "kind", "mode", "unit"]);
        if raw.keys().map(String::as_str).collect::<BTreeSet<_>>() != expected {
            return Err(config_error(format!(
                "resource binding {name} must contain exactly backend, kind, mode, and unit"
            )));
        }
        let kind = raw["kind"]
            .as_str()
            .ok_or_else(|| config_error(format!("resource binding {name} kind is invalid")))?;
        let unit = raw["unit"]
            .as_str()
            .ok_or_else(|| config_error(format!("resource binding {name} unit is invalid")))?;
        if !valid_kind_unit(kind, unit) {
            return Err(config_error(format!(
                "resource binding {name} has an unsupported kind and unit"
            )));
        }
        let mode = raw["mode"]
            .as_str()
            .filter(|mode| matches!(*mode, "admission-only" | "best-effort" | "required"))
            .ok_or_else(|| config_error(format!("resource binding {name} mode is invalid")))?;
        let backend = match raw.get("backend") {
            Some(Value::Null) => None,
            Some(Value::String(backend)) if name_valid(backend) => Some(backend.clone()),
            _ => {
                return Err(config_error(format!(
                    "resource binding {name} backend is invalid"
                )));
            }
        };
        if (mode == "admission-only") != backend.is_none() {
            return Err(config_error(format!(
                "resource binding {name} backend does not match its mode"
            )));
        }
        selected.insert(
            name.clone(),
            Binding {
                backend,
                kind: kind.to_owned(),
                mode: mode.to_owned(),
                unit: unit.to_owned(),
            },
        );
    }
    Ok(selected)
}

pub fn bindings_value(bindings: &BTreeMap<String, Binding>) -> Value {
    Value::Object(
        bindings
            .iter()
            .map(|(name, binding)| (name.clone(), binding.to_value()))
            .collect(),
    )
}

pub fn parse_bindings_json(value: &str) -> Result<BTreeMap<String, Binding>> {
    let value: Value = serde_json::from_str(value).map_err(|_| {
        AppError::new(
            "broker-owner-metadata-invalid",
            "live owner resource bindings are not valid JSON",
        )
    })?;
    parse_bindings(Some(&value)).map_err(|_| {
        AppError::new(
            "broker-owner-metadata-invalid",
            "live owner resource bindings are invalid",
        )
    })
}

pub fn load_configuration(state_dir: &Path) -> Result<ResourceConfiguration> {
    let path = state_dir.join("config.json");
    let text = match fs::read_to_string(&path) {
        Ok(text) => text,
        Err(error) if error.kind() == std::io::ErrorKind::NotFound => {
            return Ok(ResourceConfiguration {
                capacities: BTreeMap::from([("jobs".to_owned(), 2)]),
                bindings: BTreeMap::new(),
                cgroup_root: None,
                cgroup_io_paths: Vec::new(),
            });
        }
        Err(error) => {
            return Err(config_error(format!(
                "cannot read broker configuration: {error}"
            )));
        }
    };
    let document: Value = serde_json::from_str(&text)
        .map_err(|_| config_error("broker configuration is not valid JSON"))?;
    let object = document
        .as_object()
        .ok_or_else(|| config_error("broker configuration must be one JSON object"))?;
    let mut capacities = object
        .get("capacities")
        .map(|value| {
            value
                .as_object()
                .ok_or_else(|| config_error("broker configuration capacities must be an object"))?
                .iter()
                .map(|(name, units)| {
                    if !name_valid(name) {
                        return Err(config_error("capacity name is invalid"));
                    }
                    let units = units
                        .as_u64()
                        .filter(|units| *units > 0 && *units <= i64::MAX as u64)
                        .ok_or_else(|| config_error(format!("capacity {name} is invalid")))?;
                    Ok((name.clone(), units))
                })
                .collect::<Result<BTreeMap<_, _>>>()
        })
        .transpose()?
        .unwrap_or_default();
    capacities.entry("jobs".to_owned()).or_insert(2);
    let bindings = parse_bindings(object.get("bindings"))?;
    let cgroup_root = object
        .get("cgroup_root")
        .and_then(Value::as_str)
        .map(PathBuf::from)
        .map(|root| {
            std::path::absolute(root)
                .map_err(|_| config_error("cannot resolve configured cgroup root"))
        })
        .transpose()?;
    let cgroup_io_paths = object
        .get("cgroup_io")
        .and_then(Value::as_object)
        .and_then(|section| section.get("paths"))
        .and_then(Value::as_array)
        .map(|paths| {
            paths
                .iter()
                .filter_map(Value::as_str)
                .map(PathBuf::from)
                .collect()
        })
        .unwrap_or_default();
    Ok(ResourceConfiguration {
        capacities,
        bindings,
        cgroup_root,
        cgroup_io_paths,
    })
}

pub fn resource_contract(
    resources: &BTreeMap<String, u64>,
    bindings: &BTreeMap<String, Binding>,
) -> Result<Value> {
    let mut contract = Map::new();
    let mut scratch_kinds = BTreeSet::new();
    for name in resources.keys() {
        let binding = bindings
            .get(name)
            .cloned()
            .unwrap_or_else(Binding::admission);
        if binding.enforced() && matches!(binding.kind.as_str(), "storage" | "tmpfs") {
            scratch_kinds.insert(binding.kind.clone());
        }
        contract.insert(name.clone(), binding.to_value());
    }
    if scratch_kinds.len() > 1 {
        return Err(AppError::new(
            "broker-submission-invalid",
            "one run cannot combine persistent-storage and tmpfs scratch providers",
        ));
    }
    Ok(Value::Object(contract))
}

pub fn initial_receipt(resources: &BTreeMap<String, u64>) -> Value {
    json!({
        "requested": resources,
        "applied": {},
        "peak": {},
        "events": [],
    })
}

pub fn receipt_valid(value: &Value, resources: &BTreeMap<String, u64>) -> bool {
    let Some(receipt) = value.as_object() else {
        return false;
    };
    if receipt.keys().map(String::as_str).collect::<BTreeSet<_>>()
        != BTreeSet::from(["applied", "events", "peak", "requested"])
    {
        return false;
    }
    let mapping = |name: &str, allow_zero: bool| -> Option<BTreeMap<String, u64>> {
        receipt
            .get(name)?
            .as_object()?
            .iter()
            .map(|(resource, raw)| {
                raw.as_u64()
                    .filter(|units| allow_zero || *units > 0)
                    .map(|units| (resource.clone(), units))
            })
            .collect()
    };
    let Some(requested) = mapping("requested", false) else {
        return false;
    };
    let Some(applied) = mapping("applied", false) else {
        return false;
    };
    let Some(peak) = mapping("peak", true) else {
        return false;
    };
    if &requested != resources
        || applied
            .iter()
            .any(|(name, units)| requested.get(name).is_none_or(|limit| units > limit))
        || peak.keys().any(|name| !requested.contains_key(name))
    {
        return false;
    }
    let stages = BTreeSet::from([
        "probe", "prepare", "attach", "usage", "finish", "cancel", "cleanup",
    ]);
    let statuses = BTreeSet::from(["applied", "recorded", "unapplied", "failed"]);
    receipt
        .get("events")
        .and_then(Value::as_array)
        .is_some_and(|events| {
            events.iter().all(|event| {
                let Some(event) = event.as_object() else {
                    return false;
                };
                event.keys().map(String::as_str).collect::<BTreeSet<_>>()
                    == BTreeSet::from(["at", "backend", "code", "resource", "stage", "status"])
                    && event
                        .get("at")
                        .and_then(Value::as_str)
                        .is_some_and(|value| !value.is_empty())
                    && event
                        .get("backend")
                        .and_then(Value::as_str)
                        .is_some_and(name_valid)
                    && event
                        .get("resource")
                        .and_then(Value::as_str)
                        .is_some_and(|name| requested.contains_key(name))
                    && event
                        .get("stage")
                        .and_then(Value::as_str)
                        .is_some_and(|stage| stages.contains(stage))
                    && event
                        .get("status")
                        .and_then(Value::as_str)
                        .is_some_and(|status| statuses.contains(status))
                    && event
                        .get("code")
                        .and_then(Value::as_str)
                        .is_some_and(code_valid)
            })
        })
}

pub fn parse_capabilities_json(value: &str) -> Result<Value> {
    let value: Value = serde_json::from_str(value).map_err(|_| {
        AppError::new(
            "broker-owner-metadata-invalid",
            "live owner resource capabilities are not valid JSON",
        )
    })?;
    if !value.is_object() {
        return Err(AppError::new(
            "broker-owner-metadata-invalid",
            "live owner resource capabilities are invalid",
        ));
    }
    Ok(value)
}

pub fn capability_issue(binding: &Binding, capability: Option<&Value>) -> Option<String> {
    let Some(capability) = capability.and_then(Value::as_object) else {
        return Some("backend-unavailable".to_owned());
    };
    if capability.get("available").and_then(Value::as_bool) != Some(true) {
        return capability
            .get("reason")
            .and_then(Value::as_str)
            .filter(|reason| code_valid(reason))
            .map(str::to_owned)
            .or_else(|| Some("backend-unavailable".to_owned()));
    }
    let contains = |field: &str, expected: &str| {
        capability
            .get(field)
            .and_then(Value::as_array)
            .is_some_and(|values| values.iter().any(|value| value.as_str() == Some(expected)))
    };
    if !contains("kinds", &binding.kind) {
        return Some("kind-unsupported".to_owned());
    }
    if !contains("units", &binding.unit) {
        return Some("unit-unsupported".to_owned());
    }
    let lifecycle = RESOURCE_OPERATIONS
        .iter()
        .all(|operation| contains("operations", operation));
    (!lifecycle).then(|| "lifecycle-unsupported".to_owned())
}
